File: UnitTestCSharp/TestGeoCoordinateBoundingBox2d.cs
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using QuadKeys;

namespace QuadKeyTests
{
    [TestClass]
    public class TestGeoCoordinateBoundingBox2dTest
    {
        [TestMethod]
        public void TestDefaultCtor()
        {
            GeoCoordinateBoundingBox2d empty =
                new GeoCoordinateBoundingBox2d();
            Assert.IsFalse(empty.isValid());
            Assert.IsFalse(empty.contains(
                new GeoCoordinate2d(new Longitude(0.0), new Latitude(0.0))));
        }

        [TestMethod]
        public void TestCtorStandardConstruction()
        {
            { // Lower Left / Upper Right Construction
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(0.0)),
                        new GeoCoordinate2d(
                            new Longitude(180.0), new Latitude(90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(90.0)).EqualTo(geoBox.upperRight()));
                Assert.IsTrue(geoBox.isValid());
            }

            { // Upper Left / Lower Right Construction
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(90.0)),
                        new GeoCoordinate2d(
                            new Longitude(180.0), new Latitude(0.0)));

                Assert.IsTrue(new GeoCoordinate2d(
                    new Longitude(0.0), new Latitude(0.0)).EqualTo(
                    geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(
                    new Longitude(180.0), new Latitude(90.0)).EqualTo(
                    geoBox.upperRight()));
                Assert.IsTrue(geoBox.isValid());
            }

            { // Upper Right / Lower Left Construction
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(180.0), new Latitude(0.0)),
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(90.0)).EqualTo(geoBox.upperRight()));
                Assert.IsTrue(geoBox.isValid());
            }

            { // Lower Right / Upper Left Construction
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(180.0), new Latitude(0.0)),
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(90.0)).EqualTo(geoBox.upperRight()));
                Assert.IsTrue(geoBox.isValid());
            }
        }

        [TestMethod]
        public void TestCtorLinearBoxAndPoint()
        {
            { // Horizontal
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(0.0)),
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(90.0)).EqualTo(geoBox.upperRight()));
                Assert.IsTrue(geoBox.isValid());
            }

            { // Vertical
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(0.0)),
                        new GeoCoordinate2d(
                            new Longitude(180.0), new Latitude(0.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(0.0)).EqualTo(geoBox.upperRight()));
                Assert.IsTrue(geoBox.isValid());
            }

            { // Point
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(0.0)),
                        new GeoCoordinate2d(
                            new Longitude(0.0), new Latitude(0.0)));
                Assert.IsTrue(geoBox.isValid());
            }
        }

        [TestMethod]
        public void TestContainsOrNot()
        {
            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)),
                        new GeoCoordinate2d(
                            new Longitude(-90.0), new Latitude(-45.0)));

                Assert.IsTrue(geoBox.isValid());

                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(0.0), new Latitude(0.0))));

                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.0), new Latitude(45.0))));
                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.0), new Latitude(-45.0))));
                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(-90.0), new Latitude(-45.0))));
                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(-90.0), new Latitude(45.0))));

                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.01), new Latitude(45.01))));
                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.01), new Latitude(-45.01))));
                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(-90.01), new Latitude(-45.01))));
                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(-90.01), new Latitude(45.01))));
            }

            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)),
                        new GeoCoordinate2d(
                            new Longitude(0.1), new Latitude(0.1)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(0.1),
                    new Latitude(0.1)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));

                Assert.IsTrue(geoBox.isValid());

                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.0), new Latitude(45.0))));
                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.0), new Latitude(0.1))));
                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(0.1), new Latitude(0.1))));
                Assert.IsTrue(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(0.1), new Latitude(45.0))));

                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.01), new Latitude(45.01))));
                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(90.01), new Latitude(0.0))));
                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(0.0), new Latitude(0.0))));
                Assert.IsFalse(geoBox.contains(new GeoCoordinate2d(
                    new Longitude(0.0), new Latitude(45.01))));
            }
        }

        [TestMethod]
        public void TestGrowToContain()
        {
            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(-90.0), new Latitude(-45.0)),
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));

                geoBox.growToContain(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(90.0)).EqualTo(geoBox.upperRight()));
            }

            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(-90.0), new Latitude(-45.0)),
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));

                geoBox.growToContain(new GeoCoordinate2d(new Longitude(-180.0),
                    new Latitude(-90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-180.0),
                    new Latitude(-90.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));
            }

            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(-90.0), new Latitude(-45.0)),
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));

                geoBox.growToContain(new GeoCoordinate2d(new Longitude(-180.0),
                    new Latitude(90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-180.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(90.0)).EqualTo(geoBox.upperRight()));
            }

            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(-90.0), new Latitude(-45.0)),
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));

                geoBox.growToContain(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(-90.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-90.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(180.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));
            }

            {
                GeoCoordinateBoundingBox2d geoBox =
                    new GeoCoordinateBoundingBox2d(
                        new GeoCoordinate2d(
                            new Longitude(-90.0), new Latitude(-45.0)),
                        new GeoCoordinate2d(
                            new Longitude(90.0), new Latitude(45.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));

                geoBox.growToContain(new GeoCoordinate2d(new Longitude(0.0),
                    new Latitude(0.0)));

                Assert.IsTrue(new GeoCoordinate2d(new Longitude(-90.0),
                    new Latitude(-45.0)).EqualTo(geoBox.lowerLeft()));
                Assert.IsTrue(new GeoCoordinate2d(new Longitude(90.0),
                    new Latitude(45.0)).EqualTo(geoBox.upperRight()));
            }
        }
    }
}

File: test/test_geo_coordinate_bounding_box2d.cpp
#include <gtest/gtest.h>

#include "quad_keys/geo_coordinate_bounding_box2d.h"
#include "quad_keys/geo_coordinate2d.h"

#include <iostream>

namespace quad_keys
{
  class TestGeoCoordinateBoundingBox2d : public ::testing::Test
  {
  protected:
    virtual void SetUp() {}
    virtual void TearDown() {}
  };

  TEST(GEOCOORDINATEBOUNDINGBOX2D, default_ctor)
  {
    GeoCoordinateBoundingBox2d empty;
    ASSERT_FALSE(empty.is_valid());
    ASSERT_FALSE(empty.contains(
      GeoCoordinate2d(Longitude(0.0), Latitude(0.0))));
  }

  TEST(GEOCOORDINATEBOUNDINGBOX2D, ctor_standard_construction)
  {
    { // Lower Left / Upper Right Construction
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        GeoCoordinate2d(Longitude(180.0), Latitude(90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
        geo_box.upper_right());
      EXPECT_TRUE(geo_box.is_valid());
    }

    { // Upper Left / Lower Right Construction
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(0.0), Latitude(90.0)),
        GeoCoordinate2d(Longitude(180.0), Latitude(0.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
        geo_box.upper_right());
      EXPECT_TRUE(geo_box.is_valid());
    }

    { // Upper Right / Lower Left Construction
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(180.0), Latitude(0.0)),
        GeoCoordinate2d(Longitude(0.0), Latitude(90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
        geo_box.upper_right());
      EXPECT_TRUE(geo_box.is_valid());
    }

    { // Lower Right / Upper Left Construction
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(180.0), Latitude(0.0)),
        GeoCoordinate2d(Longitude(0.0), Latitude(90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
        geo_box.upper_right());
      EXPECT_TRUE(geo_box.is_valid());
    }
  }

  TEST(GEOCOORDINATEBOUNDINGBOX2D, ctor_linear_box_and_point)
  {
    { // Horizontal
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        GeoCoordinate2d(Longitude(0.0), Latitude(90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(90.0)),
        geo_box.upper_right());
      EXPECT_TRUE(geo_box.is_valid());
    }

    { // Vertical
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        GeoCoordinate2d(Longitude(180.0), Latitude(0.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(0.0)),
        geo_box.upper_right());
      EXPECT_TRUE(geo_box.is_valid());
    }

    { // Point
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(0.0), Latitude(0.0)),
        GeoCoordinate2d(Longitude(0.0), Latitude(0.0)));

      EXPECT_TRUE(geo_box.is_valid());
    }
  }

  TEST(GEOCOORDINATEBOUNDINGBOX2D, contains_or_not)
  {
    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)));

      EXPECT_TRUE(geo_box.is_valid());

      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(0.0),
        Latitude(0.0))));

      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(90.0),
        Latitude(45.0))));
      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(90.0),
        Latitude(-45.0))));
      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(-90.0),
        Latitude(-45.0))));
      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(-90.0),
        Latitude(45.0))));

      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(90.01),
        Latitude(45.01))));
      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(90.01),
        Latitude(-45.01))));
      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(-90.01),
        Latitude(-45.01))));
      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(-90.01),
        Latitude(45.01))));
    }

    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        GeoCoordinate2d(Longitude(0.1), Latitude(0.1)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(0.1), Latitude(0.1)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());

      EXPECT_TRUE(geo_box.is_valid());

      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(90.0),
        Latitude(45.0))));
      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(90.0),
        Latitude(0.1))));
      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(0.1),
        Latitude(0.1))));
      EXPECT_TRUE(geo_box.contains(GeoCoordinate2d(Longitude(0.1),
        Latitude(45.0))));

      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(90.01),
        Latitude(45.01))));
      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(90.01),
        Latitude(0.0))));
      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(0.0),
        Latitude(0.0))));
      EXPECT_FALSE(geo_box.contains(GeoCoordinate2d(Longitude(0.0),
        Latitude(45.01))));
    }
  }

  TEST(GEOCOORDINATEBOUNDINGBOX2D, grow_to_contain)
  {
    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());

      geo_box.grow_to_contain(GeoCoordinate2d(Longitude(180.0),
        Latitude(90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(90.0)),
        geo_box.upper_right());
    }

    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());

      geo_box.grow_to_contain(GeoCoordinate2d(Longitude(-180.0),
        Latitude(-90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-180.0), Latitude(-90.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());
    }

    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());

      geo_box.grow_to_contain(GeoCoordinate2d(Longitude(-180.0),
        Latitude(90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-180.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(90.0)),
        geo_box.upper_right());
    }

    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());

      geo_box.grow_to_contain(GeoCoordinate2d(Longitude(180.0),
        Latitude(-90.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-90.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(180.0), Latitude(45.0)),
        geo_box.upper_right());
    }

    {
      GeoCoordinateBoundingBox2d geo_box(
        GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        GeoCoordinate2d(Longitude(90.0), Latitude(45.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());

      geo_box.grow_to_contain(GeoCoordinate2d(Longitude(0.0),
        Latitude(0.0)));

      EXPECT_EQ(GeoCoordinate2d(Longitude(-90.0), Latitude(-45.0)),
        geo_box.lower_left());
      EXPECT_EQ(GeoCoordinate2d(Longitude(90.0), Latitude(45.0)),
        geo_box.upper_right());
    }
  }
}
